runtime/tests/interpreter/interpreter_test.go

```diff
@@ -11295,6 +11295,225 @@ func TestInterpretArrayToVariableSized(t *testing.T) {
+				),
+			),
+		)
+	})
+
+	t.Run("with string array", func(t *testing.T) {
+		t.Parallel()
+
+		inter := parseCheckAndInterpret(t, `
+			let xs: [String] = ["abc", "def"]
+
+			fun test(): [String; 2]? {
+				return xs.toConstantSized<[String; 2]>()
+			}
+		`)
+
+		runValidCase(
+			t,
+			inter,
+			interpreter.NewSomeValueNonCopying(
+				inter,
+				interpreter.NewArrayValue(
+					inter,
+					interpreter.EmptyLocationRange,
+					&interpreter.ConstantSizedStaticType{
+						Type: interpreter.PrimitiveStaticTypeString,
+						Size: 2,
+					},
+					common.ZeroAddress,
+					interpreter.NewUnmeteredStringValue("abc"),
+					interpreter.NewUnmeteredStringValue("def"),
+				),
+			),
+		)
+	})
+
+	t.Run("with wrong size", func(t *testing.T) {
+		t.Parallel()
+
+		inter := parseCheckAndInterpret(t, `
+			let xs: [Int] = [1, 2, 3, 100, 201]
+
+			fun test(): [Int; 4]? {
+				return xs.toConstantSized<[Int; 4]>()
+			}
+		`)
+
+		runValidCase(
+			t,
+			inter,
+			interpreter.NilOptionalValue,
+		)
+	})
+
+	t.Run("with array of struct", func(t *testing.T) {
+		t.Parallel()
+
+		inter := parseCheckAndInterpret(t, `
+			struct TestStruct {
+				var test: Int
+
+				init(_ t: Int) {
+					self.test = t
+				}
+			}
+
+			let sa: [TestStruct] = [TestStruct(1), TestStruct(2), TestStruct(3)]
+
+			fun test(): [TestStruct;3]? {
+				return sa.toConstantSized<[TestStruct;3]>()
+			}
+		`)
+
+		location := common.Location(common.StringLocation("test"))
+		value1 := interpreter.NewCompositeValue(
+			inter,
+			interpreter.EmptyLocationRange,
+			location,
+			"TestStruct",
+			common.CompositeKindStructure,
+			[]interpreter.CompositeField{
+				{
+					Name:  "test",
+					Value: interpreter.NewUnmeteredIntValueFromInt64(1),
+				},
+			},
+			common.ZeroAddress,
+		)
+		value2 := interpreter.NewCompositeValue(
+			inter,
+			interpreter.EmptyLocationRange,
+			location,
+			"TestStruct",
+			common.CompositeKindStructure,
+			[]interpreter.CompositeField{
+				{
+					Name:  "test",
+					Value: interpreter.NewUnmeteredIntValueFromInt64(2),
+				},
+			},
+			common.ZeroAddress,
+		)
+		value3 := interpreter.NewCompositeValue(
+			inter,
+			interpreter.EmptyLocationRange,
+			location,
+			"TestStruct",
+			common.CompositeKindStructure,
+			[]interpreter.CompositeField{
+				{
+					Name:  "test",
+					Value: interpreter.NewUnmeteredIntValueFromInt64(3),
+				},
+			},
+			common.ZeroAddress,
+		)
+
+		runValidCase(
+			t,
+			inter,
+			interpreter.NewSomeValueNonCopying(
+				inter,
+				interpreter.NewArrayValue(
+					inter,
+					interpreter.EmptyLocationRange,
+					&interpreter.ConstantSizedStaticType{
+						Type: interpreter.NewCompositeStaticType(
+							nil,
+							common.Location(common.StringLocation("test")),
+							"TestStruct",
+							"S.test.TestStruct",
+						),
+						Size: 3,
+					},
+					common.ZeroAddress,
+					value1,
+					value2,
+					value3,
+				),
+			),
+		)
+	})
+}
```

Comment:
I think this also work on array references, correct? Can you maybe add a test?

I didn't realize this earlier: but since we already have `toVariableSized` function, with the introduction of this, one can get a dereferenced copy of an array, by going the round trip. The catch here is  dereferenced-copy feature doesn't work for all types of arrays, where as this would work for all non-resource typed arrays

Reply:
Having `toVariableSized` alone itself is also not ideal, it in a way, dereferences the array.
We need a way to restrict certain functions from being called on references.

Reply:
Sorry, I'm not quite following here, what is the problem?

Reply:
The issue here is, the dereference operator only works for a subset of AnyStruct types (only primitives and collection of primitives).
But one can bypass that restriction by using these two methods, and 'dereference' an array of AnyStruct :
```cadence
fun dereference(arrayRef &[AnyStruct]): [AnyStruct] {
    constantSized = arrayRef.toConstantSized<AnyStruct; n>()
    return constantSized.toVariableSized()
}
```

Reply:
The dereferencing isn't recursive, so that shouldn't be a problem, no?

Reply:
So based on the discussion here: https://discord.com/channels/613813861610684416/1108479699732152503/1197977922057814117, this should be OK. We may instead update the dereference operator to be consistent.

Reply:
Opened to adjust the reference operator: https://github.com/onflow/cadence/pull/3034